src/models/event.js

```diff
@@ -340,19 +340,34 @@ utils.extend(module.exports.MatrixEvent.prototype, {
      * @internal
      *
      * @param {module:crypto} crypto crypto module
+     * @param {boolean} cancelAndRetry retry the attempt, even if we tried to decrypt
+     *                                 this event before. If there's an existing
+     *                                 pending key request for this event, cancel it
+     *                                 before retrying.
      *
      * @returns {Promise} promise which resolves (to undefined) when the decryption
      * attempt is completed.
      */
-    attemptDecryption: async function(crypto) {
+    attemptDecryption: async function(crypto, cancelAndRetry=false) {
         // start with a couple of sanity checks.
         if (!this.isEncrypted()) {
             throw new Error("Attempt to decrypt event which isn't encrypted");
         }
 
-        if (
-            this._clearEvent && this._clearEvent.content &&
+        if (cancelAndRetry) {
```

Comment:
what happens if we get here while the previous decryption attempt is still in progress? (ie `_decryptionPromise` is truthy)

Reply:
If the decryption is ongoing (attempt A) and we call `attemptDecryption` with `cancelAndRetry = true` (attempt B), then we cancel any ongoing key request and return `_decryptionPromise`. 

So either:
 - We attempted decryption (A), were successful and therefore didn't send a key request. We then cancel the key request (via B), but none was present (nop). Finally we return the `_decryptionPromise`.
 - We attempted decryption (A), failed and therefore sent a key request. We then cancel the key request (via B), and either:
     - It was in the UNSENT or SENT state, in which case we cancel it or 
     - It was in CANCELLATION_PENDING state somehow (nop) or
     - we throw, but this is swallowed.
 - We then continue to try and decrypt the event, which will fail causing a new key request to be sent.

So it's not optimal. If decryption is ongoing, we probably shouldn't cancel any requests.

Reply:
presumably this is quite a likely scenario, as an overenthusiastic user will stab the fudge button on adjacent messages (which is doubly bad if we end up sending multiple keyshare requests for the same session?)